[receiver/vcenterreceiver] Optimize vCenter VM network calls

Caleb-Hurshman opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
Changes the method for collecting VMs used by the `vccenterreceiver` to the more time-efficient `CreateContainerView` method. This is the first step to addressing the issue linked below.

**Link to tracking Issue:** #31837

**Testing:** These changes were tested on an environment with 200+ virtual machines. The original collection time was ~80 seconds. Collection times with these changes are ~40 seconds.

**Documentation:** N/A

Caleb-Hurshman committed:
change vm collection method

Caleb-Hurshman committed:
remove destroy ctx

Caleb-Hurshman committed:
fix client test

StefanKurek commented on .chloggen/feat_faster-vcenter-vm-network-calls.yaml:
```diff
@@ -0,0 +1,27 @@
+component: vcenterreceiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: "Changes the method for collecting VMs used by the `vccenterreceiver` to the more time efficient `CreateContainerView` method."
```
I would modify this a bit. really we are using a container view's "retrieve" method now

StefanKurek commented on receiver/vcenterreceiver/scraper.go:
```diff
@@ -306,32 +306,25 @@ func (v *vcenterMetricScraper) collectVMs(
-			errs.AddPartial(1, err)
+		hostRef := vm.Summary.Runtime.Host
+		if hostRef == nil {
+			errs.AddPartial(1, errors.New("VM doesn't have a HostSystem"))
```
This should actually be more specific and include what VM has this issue.

StefanKurek commented on receiver/vcenterreceiver/scraper.go:
```diff
@@ -349,12 +342,6 @@ func (v *vcenterMetricScraper) collectVMs(
 			}
 		}
 
-		hostname, err := vmHost.ObjectName(ctx)
```
minor and wouldn't hold up PR: since hostname is references a few times, it's still probably worth it to make a variable out of it after creating the `hwSum` variable

StefanKurek reviewed (changes requested):
From what I can see, there are fewer VMs being returned than before with this logic. Did you manually check this?

StefanKurek commented:
> From what I can see, there are fewer VMs being returned than before with this logic. Did you manually check this?

@Caleb-Hurshman ignore this comment. I was looking at a different metric which only returned for active VMs

StefanKurek commented:
> From what I can see, there are fewer VMs being returned than before with this logic. Did you manually check this?

@Caleb-Hurshman ignore this. I was looking at a wrong metric which only reported when VMs were active

Caleb-Hurshman committed:
pr feedback tweaks

StefanKurek committed:
vcenterreceiver small refactors & single perf metric network call

Makes a few small refactors for the vcenterreceiver scraper and metrics files.
Fixes unit tests.
Makes a single performance metrics network call for VMs rather than one per VM.

StefanKurek commented:
With a few additional changes, collection times in our environment dropped from 90s to 23s-30s

djaglowski commented on receiver/vcenterreceiver/testdata/metrics/expected.yaml:
```diff
@@ -2636,6 +2636,12 @@ resourceMetrics:
         - key: vcenter.vm.name
           value:
             stringValue: CentOS 7
+        - key: vcenter.resource_pool.name
+          value:
+            stringValue: Resources
+        - key: vcenter.resource_pool.inventory_path
+          value:
+            stringValue: /Datacenter/host/Cluster/Resources
```
Is there a bug fix or other change here, in addition to the optimization? If not, why would the expected result change?

StefanKurek replied:
@djaglowski I think I would argue that I improved the tests here by having better XML responses being returned (which resulted in updated expected test results). Now the tests show the new resource pool attributes being added to VMs.

djaglowski replied:
Thanks for explaining. Would it be possible to submit the changes to the mock server separately? I'm sure they're good changes but it would be helpful to validate the refactoring independently.

StefanKurek replied:
@djaglowski I can do this. 

I do want to call out that the reason that these changes are there are that I think they pertain to the testing of this block of work (the VM code). 

I made changes, the tests were broken.
So I started to update the tests, with new recordings from my new requests. But then I noticed that the expected results didn't seem to match what the code was doing (which was what I WAS expecting....if that makes any sense 😄 ).

Like I said, I can work around this and modify this PR to remove the expected results modifications. But it will make the tests in a less than optimal state IMO.

djaglowski replied:
Will the test changes work without the other changes in this PR? I was thinking that we would first update tests, then rebase this PR so it is only refactoring for performance.

StefanKurek replied:
@djaglowski let me give that a try. I think that should work and not be too difficult to implement. 👍

StefanKurek replied:
@djaglowski rebased off the main which contained the merged in fixes from the previous PR

StefanKurek committed:
vcenterreceiver remove mockserver responses that are no longer in use

StefanKurek committed:
Merge branch 'main' into feat/faster-vcenter-vm-network-calls
